devolutions-gateway/src/api/network_scan.rs

```diff
@@ -0,0 +1,123 @@
+use std::net::IpAddr;
+
+use axum::{
+    extract::{ws::Message, State, WebSocketUpgrade},
+    response::Response,
+};
+
+use network_scanner::scanner::NetworkScannerParams;
+use serde::Serialize;
+
+use crate::{http::HttpError, DgwState};
+
+pub async fn handler(
+    _: State<DgwState>,
+    ws: WebSocketUpgrade,
+    query_params: axum::extract::Query<NetworkScanQueryParams>,
+) -> Result<Response, HttpError> {
+    let scanner_params: NetworkScannerParams = query_params.0.into();
+
+    let scanner = network_scanner::scanner::NetworkScanner::new(scanner_params).map_err(|e| {
+        tracing::error!("Failed to create network scanner: {:?}", e);
+        HttpError::internal().build(e)
+    })?;
+
+    let res = ws.on_upgrade(move |mut websocket| async move {
+        let stream = scanner.start().expect("Failed to start network scanner");
+        tracing::info!("Network scan started");
+        loop {
+            tokio::select! {
+                result = stream.recv() => {
+                    let Some((ip, dns, port)) = result else{
+                        break;
+                    };
+                    let response = NetworkScanResponse::new(ip, port, dns);
+                    let Ok(response) = serde_json::to_string(&response) else {
+                        warn!("Failed to serialize response");
+                        continue;
+                    };
+
+                    if let Err(e) = websocket.send(Message::Text(response)).await {
+                        tracing::info!("Failed to send message: {:?}", e);
+                        break;
+                    }
+
+                },
+                msg = websocket.recv() => {
+
+                    let Some(msg) = msg else {
+                        break;
+                    };
+
+                    if let Ok(Message::Close(_)) = msg {
+                        break;
+                    }
+                }
+            }
+        }
+        tracing::info!("Network scan finished");
```

Comment:
nitpick: tracing macros are all imported by default (see the #[macro_use] in lib.rs). Same at the other places.

```suggestion
        info!("Network scan finished");
```

Reply:
updated